util/eventbus/eventbustest/eventbustest.go

```diff
@@ -0,0 +1,229 @@
+type Watcher struct {
+	mon    *eventbus.Subscriber[eventbus.RoutedEvent]
+	events chan any
+	chDone chan bool
+	// TimeOut defines when the Expect* functions should stop looking for events
+	// coming from the Watcher. The value is set by [NewWatcher] and defaults to
+	// the deadline passed in by [testing.T]. If looking to verify the absence
+	// of an event, the TimeOut can be set to a lower value after creating the
+	// Watcher.
+	TimeOut time.Duration
+}
+
+// Expect is a particular implementation of [ExpectFunc] that tests for the
+// existence of an event without caring about the contents of it.
+func Expect[T any](tw *Watcher) error {
+	return ExpectFunc(tw, func(event T) (bool, error) { return true, nil })
+}
+
+// ExpectFunc tests for a particular event but also a particular shape of said event.
+// This allows for looking for a single event with a specific value, or a
+// particular event in a pile of SomeEvent.
+func ExpectFunc[T any](tw *Watcher, test func(event T) (bool, error)) error {
+	eventCount := 0
+	for {
+		select {
+		case event := <-tw.events:
+			eventCount = eventCount + 1
+			if ev, ok := event.(T); ok {
+				if ok, err := test(ev); ok {
+					return nil
+				} else if err != nil {
+					return err
+				}
+			}
+		case <-time.After(tw.TimeOut):
+			return fmt.Errorf("timed out waiting for event, saw %d events", eventCount)
+		}
+	}
+}
+
+// ExpectEvents tests for a particular set of events in an order but other events
+// can happen at the same time. The contents of the events will not be tested,
+// only the event type itself.
+func ExpectEvents(tw *Watcher, events ...any) error {
+	eventCount := 0
+	head := 0
+	for head < len(events) {
+		select {
+		case event := <-tw.events:
+			eventCount = eventCount + 1
+			typeEvent := reflect.TypeOf(event)
+			typeHead := reflect.TypeOf(events[head])
+			if typeEvent == typeHead {
+				head = head + 1
+			}
+		case <-time.After(tw.TimeOut):
+			return fmt.Errorf(
+				"timed out waiting for event, saw %d events, %d was expected",
+				eventCount, head)
+		}
+	}
+	return nil
+}
+
+// ExpectOnlyEvents tests for a particular set of events in an order but other events
+// are not allowed to happen at the same time. The contents of the events will not be tested,
+// only the event type itself.
+func ExpectOnlyEvents(tw *Watcher, events ...any) error {
+	eventCount := 0
+	head := 0
+	for head < len(events) {
+		select {
+		case event := <-tw.events:
+			eventCount = eventCount + 1
+			typeEvent := reflect.TypeOf(event)
+			typeHead := reflect.TypeOf(events[head])
+			if typeEvent != typeHead {
+				return fmt.Errorf(
+					"expected event type %s, saw %s, at index %d",
+					typeHead, typeEvent, head)
+			}
+			head = head + 1
+		case <-time.After(tw.TimeOut):
+			return fmt.Errorf(
+				"timed out waiting for event, saw %d events, %d was expected",
+				eventCount, head)
+		}
+	}
+	return nil
+}
+
+// EventFunc is a test helper type that holds an event an a function to test the
```

Comment:
The trouble with this name is that it's convnentional to use the `*Func` naming for function types (e.g., `type CompareFunc[T any] func(a, b T) int`).

What do you think about, maybe, `EventSpec` or `Filter` or something like that?

We should also document that the `Event` field is used for its type only. Maybe we can move the details on to field comments below?